### ISSUE TITLE
[FreeRadius] Support EAP-TTLS-GTC with LDAP for hashed password storage

### ISSUE THREAD

**Issue by Ornias1993** [opnsense/plugins#2085]
**Is your feature request related to a problem? Please describe.**
Currently Ldap(s) backends are supported with FreeRadius, but only with cleartext passwords.
Cleartext passwords are a big no-go. 

Not having a way of using hashed passwords with LDAP and FreeRadius, basically makes the LDAP option unusable at best.

**Describe the solution you'd like**
It would be great if we could pick "EAP-TTLS-GTC", which would support all sorts of hashed openldap passwords, as explained here:
http://deployingradius.com/documents/protocols/compatibility.html

**Additional context**
An example config of how to setup radius with TTLS, GTC and Ldap is available here:
https://github.com/hacor/unifi-freeradius-ldap/tree/master/configs

Maintainers:
@mimugmail @fichtner

**Comment by mimugmail** [comment]
Have you configured this like with the guides already? Usually you need Samba to join a domain which is also a No-Go for firewall

**Comment by Ornias1993** [comment]
> Have you configured this like with the guides already? Usually you need Samba to join a domain which is also a No-Go for firewall

Samba has nothing to do with this at all... I dont think you know what you are talking about tbh..

**Comment by mimugmail** [comment]
No, so you are very welcome to shed some light in here. :) 
First link is just compatibility matrix which doesnt really help. Second are plain config files and I dont have openldap.

The LDAP feature is only there because some guy asked for it because he thought this might help for Wifi and I blindly integrated it. Thats why I asked if you have set this up on your own and know that this will also work with opnsense.

Dont geht me wrong but there were too many requests in the past which I added for nothing.

**Comment by Ornias1993** [comment]
> No, so you are very welcome to shed some light in here. 

Fair enough :)

> First link is just compatibility matrix which doesnt really help

It's a compatibility matrix which openldap password hash schemes are supported by which EAP modes. That does help, because you would notice your currently supported EAP modes are not supporting ANY form of hashing.

> Second are plain config files and I dont have openldap.

It's the FreeRadius config in that that maters, not the LDAP config. That explains how to setup EAP-TTLS with GTC which DOES support hashed passwords from ldap natively. (as seen in the compatibility matrix)

> The LDAP feature is only there because some guy asked for it because he thought this might help for Wifi and I blindly integrated it.

It's one of the more usefull features of FreeRadius and also quite often used in SMB too. It's not some obscuire feature as you make it out to be.

> Thats why I asked if you have set this up on your own

You didn't.

> know that this will also work with opnsense.

OPNsense FreeRadius plugin just uses normal FreeRadius config files, there is nothing that is keeping it from using a different scheme. Although it might be some work. 
Tbh, if I set it up it would be on a seperate docker container for FreeRadius and I don't have the time to also learn how to set it up on OPNSense too (including setting up a dev environment for it).

> Dont geht me wrong but there were too many requests in the past which I added for nothing.

It's not for nothing, it's kinda a requirement for current functionality to actually function as it should.

TLDR: 
It's a feature request, If I would've had the time to PR a ready-to-rock fix, it I wouldn't even have bothered filing this.

**Comment by mimugmail** [comment]
> Have you configured this like with the guides already? 

I did :)

This is only for OpenLDAP, correct? 
You don't need a dev env. I'd do it like this:

- Install FR plugin
- configure it to contact your ldap
- go to CLI and do a backup of current FR folder
- edit the files compared to the ones on your link
- stop FR via service radiusd stop and start in debug with radiusd -X
- if it's doing the things like you want if make a diff with backup folder and paste it here

This should take around 1 hour if you know a bit about FR. Without this it would take me 10x plus setup of OpenLDAP.

**Comment by Ornias1993** [comment]
> > Have you configured this like with the guides already?
> 
> I did :)

Then I misunderstoot, because there is no "guide" referenced anywhere.

> This is only for OpenLDAP, correct?
Others might want to use EAS-TTLS-GTC too, but thats not my focus, required or expected to hapen much

> You don't need a dev env. I'd do it like this

How about: I need a dev environment because i'm not going to bother users by testing this live?

> if it's doing the things like you want if make a diff with backup folder and paste it here

So TLDR: you want me to make a diff, so you can easily just add some "ifs" where needed and add the GUI handle?

**Comment by mimugmail** [comment]
Indeed, that's it, because I too have no environment to test.

**Comment by Ornias1993** [comment]
@mimugmail 
The solution, this works with both normal (user list tab) and ldap auth, but should preferably only be used with ldap:

Create an option called "EAP-TTLS-GTC", this option would change:
in `mods-enabled/eap`

Under 
`ttls {`

create an option to change:
`default_eap_type = md5`

to:
`default_eap_type = gtc`

**Thats all.**

**Comment by Ornias1993** [comment]
@mimugmail
As it's really trivial, I just submitted the PR for review :)

**Comment by mimugmail** [comment]
Please close
